File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/3f1bf7023dfd3a9069a458a0339df56f.py
class Bitset:
    def __init__(self, N):
        self.n = (N + 31) // 32
        self.data = [0] * self.n

    def set(self):
        for i in range(self.n):
            self.data[i] = 0xFFFFFFFF

    def set_bit(self, pos):
        self.data[pos >> 5] |= 1 << (pos & 31)

    def reset(self):
        for i in range(self.n):
            self.data[i] = 0

    def reset_bit(self, pos):
        self.data[pos >> 5] &= ~(1 << (pos & 31))

    def flip_bit(self, pos):
        self.data[pos >> 5] ^= 1 << (pos & 31)

    def test_bit(self, pos):
        return (self.data[pos >> 5] >> (pos & 31)) & 1

    def __or__(self, other):
        res = Bitset(len(self.data) * 32)
        for i in range(self.n):
            res.data[i] = self.data[i] | other.data[i]
        return res

    def __and__(self, other):
        res = Bitset(len(self.data) * 32)
        for i in range(self.n):
            res.data[i] = self.data[i] & other.data[i]
        return res

    def __xor__(self, other):
        res = Bitset(len(self.data) * 32)
        for i in range(self.n):
            res.data[i] = self.data[i] ^ other.data[i]
        return res

    def __lshift__(self, shift):
        res = Bitset(len(self.data) * 32)
        full = shift >> 5
        if full >= self.n:
            return res
        if full:
            res.data[full:] = self.data[:-full]
            res.data[:full] = [0] * full

        shift &= 31
        if shift:
            for i in range(len(res.data) - 1, 0, -1):
                res.data[i] = (res.data[i] << shift) | (res.data[i - 1] >> (32 - shift))
            res.data[0] <<= shift

        return res

    def __rshift__(self, shift):
        res = Bitset(len(self.data) * 32)
        full = shift >> 5
        if full >= self.n:
            return res
        if full:
            res.data[:-full] = self.data[full:]
            res.data[-full:] = [0] * full

        shift &= 31
        if shift:
            for i in range(len(res.data) - 1):
                res.data[i] = (res.data[i] >> shift) | (res.data[i + 1] << (32 - shift))
            res.data[-1] >>= shift

        return res

    def count(self):
        count = 0
        for i in range(self.n):
            count += bin(self.data[i]).count('1')
        return count

    def none(self):
        for i in range(self.n):
            if self.data[i] != 0:
                return False
        return True

    def any(self):
        for i in range(self.n):
            if self.data[i] != 0:
                return True
        return False

    def __eq__(self, other):
        return self.data == other.data

    def __lt__(self, other):
        for i in range(self.n):
            if self.data[i] != other.data[i]:
                return self.data[i] < other.data[i]
        return False

    def __gt__(self, other):
        for i in range(self.n):
            if self.data[i] != other.data[i]:
                return self.data[i] > other.data[i]
        return False

    def __le__(self, other):
        return not self.__gt__(other)

    def __ge__(self, other):
        return not self.__lt__(other)

    def __ne__(self, other):
        return not self.__eq__(other)


class IG:
    def __init__(self, n):
        self.good = [Bitset(2 * n) for _ in range(2 * n)]
        self.bad = [Bitset(2 * n) for _ in range(2 * n)]
        self.bv = [False] * (2 * n)

    @staticmethod
    def get_num(a, n):
        if a > 0:
            return a - 1
        a = -a
        return n + a - 1

    @staticmethod
    def neg(a, n):
        if a < n:
            return a + n
        return a - n

    def add_implication(self, a, b, n):
        a = self.get_num(a, n)
        b = self.get_num(b, n)
        if b < n:
            self.good[a].set_bit(b)
        else:
            self.bad[a].set_bit(b - n)

    def add(self, a, b, n):
        self.add_implication(-a, b, n)
        self.add_implication(-b, a, n)

    def transitive_closure(self, n):
        for i in range(2 * n):
            if i < n:
                self.good[i].set_bit(i)
            else:
                self.bad[i].set_bit(i - n)

        for mid in range(2 * n):
            for i in range(2 * n):
                if mid < n:
                    if self.good[i].test_bit(mid):
                        self.good[i] |= self.good[mid]
                        self.bad[i] |= self.bad[mid]
                else:
                    if self.bad[i].test_bit(mid - n):
                        self.good[i] |= self.good[mid]
                        self.bad[i] |= self.bad[mid]

        for i in range(2 * n):
            if i < n:
                if self.bad[i].test_bit(i) and self.good[i + n].test_bit(i):
                    return False
            else:
                if self.good[i].test_bit(i - n) and self.bad[i - n].test_bit(i):
                    return False

        return True

    def has_solution(self, n):
        return self.transitive_closure(n)

    def select(self, v, n):
        t = v
        if num[v] == 0:
            t += n
        if self.bv[t]:
            return False
        k = 0
        sof = [0] * n
        sof[k] = v
        ok = True
        for i in range(n):
            if self.good[t].test_bit(i):
                if num[i] == 0:
                    ok = False
                    break
                if num[i] == -1:
                    num[i] = 1
                    k += 1
                    sof[k] = i
            if self.bad[t].test_bit(i):
                if num[i] == 1:
                    ok = False
                    break
                if num[i] == -1:
                    num[i] = 0
                    k += 1
                    sof[k] = i

        if not ok:
            for i in range(k + 1):
                num[sof[i]] = -1
            return False
        return True

    def print_solution_and_die(self, f1, f2, n):
        num = [-1] * n
        if f1 != -1:
            if f1 < n:
                num[f1] = 1
            else:
                f1 -= n
                num[f1] = 0
            self.select(f1, n)
        if f2 != -1:
            if num[f2 < n ? f2 : f2 - n] == -1:
                if f2 < n:
                    num[f2] = 1
                else:
                    f2 -= n
                    num[f2] = 0
                self.select(f2, n)
        for i in range(n):
            if num[i] == -1:
                num[i] = 0
                if not self.select(i, n):
                    num[i] = 1
                    self.select(i, n)
        return num


def solve(cl1, cl2, g1, g2, m1, m2, n):
    used = [[False] * (2 * n) for _ in range(2 * n)]
    for i in range(m1):
        v1 = IG.get_num(cl1[i][0], n)
        v2 = IG.get_num(cl1[i][1], n)
        v1 = IG.neg(v1, n)
        v2 = IG.neg(v2, n)
        if v1 > v2:
            v1, v2 = v2, v1
        if used[v1][v2]:
            continue
        used[v1][v2] = True

        if g2.bv[v1] or g2.bv[v2]:
            continue
        if g2.check_bad(v1, v2):
            continue
        g2.print_solution_and_die(v1, v2, n)


n, m1, m2 = map(int, input().split())
cl1 = []
cl2 = []
for _ in range(m1):
    a, b = map(int, input().split())
    cl1.append((a, b))
for _ in range(m2):
    a, b = map(int, input().split())
    cl2.append((a, b))

g1 = IG(n)
g2 = IG(n)

for a, b in cl1:
    g1.add(a, b, n)
for a, b in cl2:
    g2.add(a, b, n)

g1.transitive_closure(n)
g2.transitive_closure(n)

r1 = g1.has_solution(n)
r2 = g2.has_solution(n)

if not r1 and not r2:
    print("SIMILAR")
    exit(0)
if not r1:
    g2.print_solution_and_die(-1, -1, n)
if not r2:
    g1.print_solution_and_die(-1, -1, n)

solve(cl1, cl2, g1, g2, m1, m2, n)
solve(cl2, cl1, g2, g1, m2, m1, n)

print("SIMILAR")

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Go/3f1bf7023dfd3a9069a458a0339df56f.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Bitset struct {
	data []uint32
}

func NewBitset(n int) *Bitset {
	data := make([]uint32, (n+31)/32)
	return &Bitset{data}
}

func (b *Bitset) Set() {
	for i := range b.data {
		b.data[i] = 0xffffffff
	}
}

func (b *Bitset) SetPos(pos int) {
	b.data[pos>>5] |= 1 << (pos & 31)
}

func (b *Bitset) Reset() {
	for i := range b.data {
		b.data[i] = 0
	}
}

func (b *Bitset) ResetPos(pos int) {
	b.data[pos>>5] &= ^(1 << (pos & 31))
}

func (b *Bitset) FlipPos(pos int) {
	b.data[pos>>5] ^= 1 << (pos & 31)
}

func (b *Bitset) TestPos(pos int) bool {
	return (b.data[pos>>5]>>(pos&31))&1 == 1
}

func (b *Bitset) Or(other *Bitset) {
	for i := range b.data {
		b.data[i] |= other.data[i]
	}
}

func (b *Bitset) And(other *Bitset) {
	for i := range b.data {
		b.data[i] &= other.data[i]
	}
}

func (b *Bitset) Xor(other *Bitset) {
	for i := range b.data {
		b.data[i] ^= other.data[i]
	}
}

func (b *Bitset) Count() int {
	count := 0
	for _, d := range b.data {
		count += popcount(d)
	}
	return count
}

func (b *Bitset) None() bool {
	for _, d := range b.data {
		if d != 0 {
			return false
		}
	}
	return true
}

func (b *Bitset) Any() bool {
	return !b.None()
}

func (b *Bitset) CountRange(l, r int) int {
	count := 0
	for i := l; i < r; i++ {
		if b.TestPos(i) {
			count++
		}
	}
	return count
}

func (b *Bitset) NoneRange(l, r int) bool {
	for i := l; i < r; i++ {
		if b.TestPos(i) {
			return false
		}
	}
	return true
}

func (b *Bitset) AnyRange(l, r int) bool {
	return !b.NoneRange(l, r)
}

func (b *Bitset) ShiftLeft(shift int) {
	full := shift >> 5
	if full >= len(b.data) {
		b.Reset()
		return
	}
	if full > 0 {
		copy(b.data[full:], b.data)
		for i := 0; i < full; i++ {
			b.data[i] = 0
		}
	}

	shift &= 31
	if shift > 0 {
		for i := len(b.data) - 1; i >= 1; i-- {
			b.data[i] = (b.data[i] << shift) | (b.data[i-1] >> (32 - shift))
		}
		b.data[0] <<= shift
	}
}

func (b *Bitset) ShiftRight(shift int) {
	full := shift >> 5
	if full >= len(b.data) {
		b.Reset()
		return
	}
	if full > 0 {
		copy(b.data, b.data[full:])
		for i := len(b.data) - full; i < len(b.data); i++ {
			b.data[i] = 0
		}
	}

	shift &= 31
	if shift > 0 {
		mask := (1 << shift) - 1
		for i := 0; i < len(b.data)-1; i++ {
			b.data[i] = (b.data[i] >> shift) | ((b.data[i+1] & mask) << (32 - shift))
		}
		b.data[len(b.data)-1] >>= shift
	}
}

func (b *Bitset) Equals(other *Bitset) bool {
	for i := range b.data {
		if b.data[i] != other.data[i] {
			return false
		}
	}
	return true
}

func popcount(x uint32) int {
	x -= (x >> 1) & 0x55555555
	x = (x & 0x33333333) + ((x >> 2) & 0x33333333)
	x = (x + (x >> 4)) & 0x0f0f0f0f
	x += x >> 8
	x += x >> 16
	return int(x & 0x0000003f)
}

type IG struct {
	good, bad []*Bitset
	bv       []bool
}

func NewIG(n int) *IG {
	good := make([]*Bitset, 2*n)
	bad := make([]*Bitset, 2*n)
	for i := range good {
		good[i] = NewBitset(2 * n)
		bad[i] = NewBitset(2 * n)
	}
	bv := make([]bool, 2*n)
	return &IG{good, bad, bv}
}

func (ig *IG) AddImplication(a, b int) {
	a = GetNum(a)
	b = GetNum(b)
	if b < n {
		ig.good[a].SetPos(b)
	} else {
		ig.bad[a].SetPos(b - n)
	}
}

func (ig *IG) Add(a, b int) {
	ig.AddImplication(-a, b)
	ig.AddImplication(-b, a)
}

func (ig *IG) TransitiveClosure() {
	for i := 0; i < 2*n; i++ {
		if i < n {
			ig.good[i].SetPos(i)
		} else {
			ig.bad[i].SetPos(i - n)
		}
	}

	for mid := 0; mid < 2*n; mid++ {
		for i := 0; i < 2*n; i++ {
			ok := false
			if mid < n {
				if ig.good[i].TestPos(mid) {
					ok = true
				}
			} else {
				if ig.bad[i].TestPos(mid - n) {
					ok = true
				}
			}
			if ok {
				ig.good[i].Or(ig.good[mid])
				ig.bad[i].Or(ig.bad[mid])
			}
		}
	}

	for i := 0; i < 2*n; i++ {
		if i < n {
			if ig.bad[i].TestPos(i) && ig.good[i+n].TestPos(i) {
				return false
			}
		} else {
			if ig.good[i].TestPos(i-n) && ig.bad[i].TestPos(i-n) {
				return false
			}
		}
	}

	for i := 0; i < 2*n; i++ {
		if ig.bv[i] {
			continue
		}
		if i < n {
			if ig.good[i].TestPos(i) {
				ig.bv[i] = true
			}
		} else {
			if ig.bad[i].TestPos(i-n) {
				ig.bv[i] = true
			}
		}
	}

	for i := 0; i < 2*n; i++ {
		if !ig.bv[i] {
			continue
		}
		for j := 0; j < 2*n; j++ {
			if !ig.bv[j] {
				continue
			}
			if j < n {
				if ig.good[i].TestPos(j) {
					ig.bv[i] = true
				}
			} else {
				if ig.bad[i].TestPos(j-n) {
					ig.bv[i] = true
				}
			}
		}
	}
	return true
}

func (ig *IG) HasSolution() bool {
	for i := 0; i < n; i++ {
		if ig.bad[i].TestPos(i) && ig.good[i+n].TestPos(i) {
			return false
		}
	}
	return true
}

func (ig *IG) Select(v int) bool {
	t := v
	if ig.num[v] == 0 {
		t += n
	}
	if ig.bv[t] {
		return false
	}
	k := 0
	ig.sof[k] = v
	ok := true
	for i := 0; i < n; i++ {
		if ig.good[t].TestPos(i) {
			if ig.num[i] == 0 {
				ok = false
				break
			}
			if ig.num[i] == -1 {
				ig.num[i] = 1
				ig.sof[k] = i
				k++
			}
		}
		if ig.bad[t].TestPos(i) {
			if ig.num[i] == 1 {
				ok = false
				break
			}
			if ig.num[i] == -1 {
				ig.num[i] = 0
				ig.sof[k] = i
				k++
			}
		}
	}

	if !ok {
		for i := 0; i < k; i++ {
			ig.num[ig.sof[i]] = -1
		}
		return false
	}
	return true
}

func (ig *IG) PrintSolutionAndDie(f1, f2 int) {
	for i := 0; i < n; i++ {
		ig.num[i] = -1
	}
	if f1 != -1 {
		if f1 < n {
			ig.num[f1] = 1
		} else {
			f1 -= n
			ig.num[f1] = 0
		}
		ig.Select(f1)
	}
	if f2 != -1 {
		if ig.num[f2 < n ? f2 : f2-n] == -1 {
			if f2 < n {
				ig.num[f2] = 1
			} else {
				f2 -= n
				ig.num[f2] = 0
			}
			ig.Select(f2)
		}
	}
	for i := 0; i < n; i++ {
		if ig.num[i] != -1 {
			continue
		}
		ig.num[i] = 0
		if !ig.Select(i) {
			ig.num[i] = 1
			ig.Select(i)
		}
	}
	for i := 0; i < n; i++ {
		if i > 0 {
			fmt.Print(" ")
		}
		fmt.Print(ig.num[i])
	}
	fmt.Println()
	os.Exit(0)
}

func (ig *IG) CheckBad(v1, v2 int) bool {
	t := ig.good[v1].Or(ig.good[v2])
	t.And(ig.bad[v1].Or(ig.bad[v2]))
	return t.Any()
}

var (
	n, m1, m2 int
	cl1, cl2  [][2]int
	g1, g2    *IG
	used      [2002][2002]bool
)

func solve(cl1, cl2 [][2]int, g1, g2 *IG, m1, m2 int) {
	for i := 0; i < m1; i++ {
		v1 := GetNum(cl1[i][0])
		v2 := GetNum(cl1[i][1])
		v1 = Neg(v1)
		v2 = Neg(v2)
		if v1 > v2 {
			v1, v2 = v2, v1
		}
		if used[v1][v2] {
			continue
		}
		used[v1][v2] = true

		if g2.bv[v1] || g2.bv[v2] {
			continue
		}
		if g2.CheckBad(v1, v2) {
			continue
		}
		g2.PrintSolutionAndDie(v1, v2)
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	n = nextInt(scanner)
	m1 = nextInt(scanner)
	m2 = nextInt(scanner)

	cl1 = make([][2]int, m1)
	cl2 = make([][2]int, m2)

	g1 = NewIG(n)
	g2 = NewIG(n)

	for i := 0; i < m1; i++ {
		a := nextInt(scanner)
		b := nextInt(scanner)
		cl1[i] = [2]int{a, b}
		g1.Add(a, b)
	}

	for i := 0; i < m2; i++ {
		a := nextInt(scanner)
		b := nextInt(scanner)
		cl2[i] = [2]int{a, b}
		g2.Add(a, b)
	}

	g1.TransitiveClosure()
	g2.TransitiveClosure()

	r1 := g1.HasSolution()
	r2 := g2.HasSolution()

	if !r1 && !r2 {
		fmt.Println("SIMILAR")
		return
	}

	if !r1 {
		g2.PrintSolutionAndDie(-1, -1)
	}

	if !r2 {
		g1.PrintSolutionAndDie(-1, -1)
	}

	solve(cl1, cl2, g1, g2, m1, m2)
	solve(cl2, cl1, g2, g1, m2, m1)

	fmt.Println("SIMILAR")
}

func nextInt(scanner *bufio.Scanner) int {
	scanner.Scan()
	ans := 0
	neg := false
	ch := scanner.Bytes()
	if ch[0] == '-' {
		neg = true
		ch = ch[1:]
	}
	for _, c := range ch {
		ans = ans*10 + int(c-'0')
	}
	if neg {
		ans = -ans
	}
	return ans
}

func GetNum(a int) int {
	if a > 0 {
		return a - 1
	}
	a = -a
	return n + a - 1
}

func Neg(a int) int {
	if a < n {
		return a + n
	}
	return a - n
}
